src/bmpcodec.c

```diff
@@ -1068,13 +1140,19 @@ gdip_read_bmp_image (void *pointer, GpImage **image, ImageSource source)
 	result->active_bitmap->scan0 = pixels;
 	result->active_bitmap->reserved = GBD_OWN_SCAN0;
 	result->active_bitmap->image_flags = ImageFlagsReadOnly | ImageFlagsHasRealPixelSize | ImageFlagsColorSpaceRGB;
-	if (bmi.biXPelsPerMeter != 0 && bmi.biYPelsPerMeter != 0)
+	if (bmi->biXPelsPerMeter != 0 && bmi->biYPelsPerMeter != 0)
 		result->active_bitmap->image_flags |= ImageFlagsHasRealDPI;
 
+	GdipFree (bmi);
+
 	*image = result;
 	return Ok;
 
 error:
+	if (bmi != NULL) {
```

Comment:
@filipnavara bmi is always deallocated here and above

Reply:
Fair enough, I missed that. I am not a big fan of functions that return failure and require additional cleanup. It would seem cleaner to me if the cleanup was done directly in gdip_read_BITMAPINFOHEADER.